core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala

```diff
@@ -236,13 +271,23 @@ class HadoopMapReduceCommitProtocol(
     committer.setupTask(taskContext)
     addedAbsPathFiles = mutable.Map[String, String]()
     partitionPaths = mutable.Set[String]()
+    stagingTaskFiles = mutable.Set[Path]()
   }
 
   override def commitTask(taskContext: TaskAttemptContext): TaskCommitMessage = {
     val attemptId = taskContext.getTaskAttemptID
     logTrace(s"Commit task ${attemptId}")
     SparkHadoopMapRedUtil.commitTask(
       committer, taskContext, attemptId.getJobID.getId, attemptId.getTaskID.getId)
+    if (dynamicPartitionOverwrite) {
+      val fs = stagingDir.getFileSystem(taskContext.getConfiguration)
+      stagingTaskFiles.foreach { stagingTaskFile =>
+        val fileName = stagingTaskFile.getName
+        val taskPartitionPath = getPartitionPath(stagingTaskFile)
+        val destFile = new Path(new Path(stagingDir, taskPartitionPath), fileName)
+        fs.rename(stagingTaskFile, destFile)
```

Comment:
My thought if in case of partition overwrite , the list of files inside a partition that needs to be modified needs to be atomically replaced.  In the above code I see that we are replacing the partitions file by file during task commit. The suggestion would be instead of rename _temporary/partitionPath/taskAttemptId/filename to partitionPath/filename, we can do rename _temporary/partitionPath/taskAttemptId/ to partitionPath/ . Please suggest.

Reply:
Thanks for your suggestion, but I think we can not do this.
For dynamic partition overwrite, each task may write data to many partition paths.
If we just rename a folder, it would overwrite other task's output.

Reply:
requesting your help in clarifying the note "If we just rename a folder, it would overwrite other task's output."  I see that we are creating _temporary path including the partition info as _temporary/partitionPath/taskAttemptId/  , if we rename this to partitionPath/ I believe it would not impact other partitions and would not overwrite the task output.